### PR TITLE
Update: `key-spacing` should allow 1+ spaces before/after colon

### LINKED ISSUE #3363: Key-spacing: allow 1+ spaces before/after colon
Currently, there doesn't seem to be a way to allow more than 1 space before/after a colon. With my teams' particular use-case, we want to _allow_ devs to align values if they wish, but have _at least_ 1 space after a colon.

### LINKED ISSUE #3453: Allow key-spacing align to align to nearest indentation level instead of a single space
The key-spacing align works great for this case:

    var x = {
        a: 1,
        b: 2
    }

But in my code, I prefer to align to the nearest indentation level (which in this case is 4-spaces). Note that there are two spaces between the colon and the value - this is because it would take two spaces to get it to the next indentation level (multiple of 4 spaces)

    var x = {
        a:  1,
        b:  2
    }

It's easier to just hit tab until things align instead of some combination of tab and spaces (in editors that support soft tabs), and it also leaves more room for objects to grow without having to modify every key value pair.

The simplest way would probably be passing in the tab size to this rule, but it would be neat if it could read the value from the indent rule.

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/25966964-allow-key-spacing-align-to-align-to-nearest-indentation-level-instead-of-a-single-space?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by BYK**
Fixes #3363.

### CODE REVIEW

**Comment by BYK** [comment]
If the approach looks good, I'll update the docs and add some more tests.

File: lib/rules/key-spacing.js [hunk 1]
```diff
@@ -333,11 +336,18 @@ module.exports.schema = [
             "align": {
                 "enum": ["colon", "value"]
             },
+            "mode": {
+                "enum": ["strict", "minimum"]
+            },
             "beforeColon": {
-                "type": "boolean"
+                "type": ["boolean", "number"],
```
Comment (btmills) [inline lib/rules/key-spacing.js]: Why do we need this to be a number as well? From the issues it sounded like `strict/minimum` would cover the use cases.

Comment (IanVS) [inline lib/rules/key-spacing.js]: I agree.  I'm not convinced about the value of forcing a particular minimum number of spaces.  Also, this wouldn't address the original issue reporter's use case of indenting to the next indentation level (total column position equaling next multiple of 2 or 4 or... well it's based on indent setting of the text editor.).

Comment (BYK) [inline lib/rules/key-spacing.js]: Okay, reverted. I guess we need a new option to pass the indent size for aligning to that. `snapTo`?

**Comment by nzakas** [comment]
Up to @btmills on this one.

File: tests/lib/rules/key-spacing.js [hunk 2]
```diff
@@ -150,6 +150,17 @@ ruleTester.run("key-spacing", rule, {
...
     }, {
+        code: [
+            "obj = { key ",
+            "    :     ",
+            " longName };"
+        ].join("\n"),
+        options: [{
+            beforeColon: true,
+            afterColon: false,
+            mode: "minimum"
```
Comment (btmills) [inline tests/lib/rules/key-spacing.js]: When `align` isn't set, the rule never warns on properties that span multiple lines - see `isSingleLine`. Putting the property key and value on the same line should validate the code to handle `minimum`.

Comment (BYK) [inline tests/lib/rules/key-spacing.js]: Added single line test.

**Comment by ilyavolodin** [comment]
I think this is good to merge now. @btmills can you take a look and merge if you agree?

**Comment by btmills** [comment]
I think the approach is good, all that's left is documenting the new option.

**Comment by BYK** [comment]
@btmills updated docs too. Ready for review.

**Pull request merged**